### ISSUE TITLE
unable to get vault provider to work

### ISSUE THREAD

**Issue by rpenugonda** [hashicorp/vault-csi-provider#95]
```
2021-05-12T18:06:58.138Z [DEBUG] server.provider: setting Vault namespace: namespace=unix
2021-05-12T18:06:58.138Z [DEBUG] server.provider: performing vault login
2021-05-12T18:06:58.138Z [DEBUG] server.provider: creating service account token bound to pod: namespace=vault serviceAccountName=webapp-sa podName=webapp podUID=34db3d17-a08b-462a-bf64-ca6c6627b3b2
2021-05-12T18:06:58.140Z [INFO]  server: Finished unary gRPC call: grpc.method=/v1alpha1.CSIDriverProvider/Mount grpc.time=2.189259ms grpc.code=Unknown err="error making mount request: failed to create a service account token for requesting pod {webapp 34db3d17-a08b-462a-bf64-ca6c6627b3b2 vault webapp-sa}: the server could not find the requested resource"
2021-05-12T18:07:06.167Z [INFO]  server: Processing unary gRPC call: grpc.method=/v1alpha1.CSIDriverProvider/Mount
2021-05-12T18:07:06.167Z [DEBUG] server: Request contents: req="attributes:"{\"csi.storage.k8s.io/pod.name\":\"webapp\",\"csi.storage.k8s.io/pod.namespace\":\"vault\",\"csi.storage.k8s.io/pod.uid\":\"34db3d17-a08b-462a-bf64-ca6c6627b3b2\",\"csi.storage.k8s.io/serviceAccount.name\":\"webapp-sa\",\"objects\":\"- objectName: \\\"db-password\\\"\\n  secretPath: \\\"concourse/unixservices/test/db-pass\\\"\\n  secretKey: \\\"password\\\"\\n\",\"roleName\":\"database\",\"vaultAddress\":\"https://vault:8200\",\"vaultKubernetesMountPath\":\"test\",\"vaultNamespace\":\"unix\",\"vaultSkipTLSVerify\":\"true\"}" secrets:"null" target_path:"/var/lib/kubelet/pods/34db3d17-a08b-462a-bf64-ca6c6627b3b2/volumes/kubernetes.io~csi/secrets-store-inline/mount" permission:"420""
```
I am seeing the above with version 0.2.0, the issue happens during the JWT creation https://github.com/hashicorp/vault-csi-provider/blob/master/internal/provider/provider.go#L45-L76.
it does have the proper clusterrole with `serviceaccounts/token` create permissions. We are using external vault and i can login to vault using the pods serviceaccount
```
vault write auth/test/login role=database jwt='xxx'
Key                                       Value
---                                       -----
token                                     s.X
token_accessor                            s.Y
token_duration                            20m
token_renewable                           true
token_policies                            ["default" "internal-app"]
identity_policies                         []
policies                                  ["default" "internal-app"]
token_meta_role                           database
token_meta_service_account_name           webapp-sa
token_meta_service_account_namespace      vault
token_meta_service_account_secret_name    webapp-sa-token-w2tqp
token_meta_service_account_uid            b1d5ff84-a3ff-4c89-90c8-93724674fe26

role contents
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: vault-csi-provider-clusterrole
rules:
- apiGroups:
  - ""
  resources:
  - serviceaccounts/token
  ```

**Comment by rpenugonda** [comment]
from audit logs where i can see the token create request and failure
```
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"d9b7a092-f3e6-4ce2-933f-c6dc7fbe61a5","stage":"RequestReceived","requestURI":"/api/v1/namespaces/vault/serviceaccounts/webapp-sa/token","verb":"create","user":{"username":"system:serviceaccount:vault:vault1-csi-provider","uid":"6f867cfb-8fb9-4e2c-b082-d31c4ebcd320","groups":["system:serviceaccounts","system:serviceaccounts:vault","system:authenticated"]},"sourceIPs":["10.250.0.104"],"userAgent":"vault-csi-provider/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"serviceaccounts","namespace":"vault","name":"webapp-sa","apiVersion":"v1","subresource":"token"},"requestReceivedTimestamp":"2021-05-12T18:17:06.968126Z","stageTimestamp":"2021-05-12T18:17:06.968126Z"}
{"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"d9b7a092-f3e6-4ce2-933f-c6dc7fbe61a5","stage":"ResponseComplete","requestURI":"/api/v1/namespaces/vault/serviceaccounts/webapp-sa/token","verb":"create","user":{"username":"system:serviceaccount:vault:vault1-csi-provider","uid":"6f867cfb-8fb9-4e2c-b082-d31c4ebcd320","groups":["system:serviceaccounts","system:serviceaccounts:vault","system:authenticated"]},"sourceIPs":["10.250.0.104"],"userAgent":"vault-csi-provider/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"serviceaccounts","namespace":"vault","name":"webapp-sa","apiVersion":"v1","subresource":"token"},"responseStatus":{"metadata":{},"status":"Failure","reason":"NotFound","code":404},"requestReceivedTimestamp":"2021-05-12T18:17:06.968126Z","stageTimestamp":"2021-05-12T18:17:06.968482Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by ClusterRoleBinding \"vault1-csi-provider-clusterrolebinding\" of ClusterRole \"vault1-csi-provider-clusterrole\" to ServiceAccount \"vault1-csi-provider/vault\""}}
```

**Comment by rpenugonda** [comment]
i am i missing the featureGate `TokenRequest` here?

**Comment by tomhjp** [comment]
Hi @rpenugonda, thanks for the detailed issue, what version of k8s are you on? The feature gate has been enabled by default since 1.12 so I think the issue here is most likely that `kube-apiserver` doesn't have the required flags for enabling the `TokenRequest` endpoint. I've never found particularly clear k8s documentation on that, but I made some notes on the flags required here: https://github.com/hashicorp/vault-csi-provider#prerequisites.

TL;DR you need to set `--service-account-signing-key-file` and `--service-account-issuer`, and `kubeadm` doesn't set those by default until 1.20, which is why I think they are most likely the missing component here.

**Comment by rpenugonda** [comment]
Thanks for the quick response. we are using 1.18.6.
after adding the 2 flags to the apisever
service-account-signing-key-file
service-account-issuer
secret store + vault works as expected. 

closing the ticket
